CIP-0138/README.md

```diff
@@ -0,0 +1,294 @@
+
+    -- again, since True does not encode to full bytes, the above example can be tricky to decode, but is nevertheless correct
+    -- 0 0000001 is the beginning of the second block, it represents the number of elements in the block which is 1
+    -- immediately after, there is 1 bit which is the encoding of the single element of this block
+    -- and finally, the last 8 zeros 
+
+</details>
+
+
+## Rationale: how does this CIP achieve its goals?
+
+### Choice of builtin functions and their specification
+
+The following section presents the reasoning behind the above specification of a Plutus
+core builtin array type.
+
+It is important to mention that we based our decisions on the desire to keep the builtin
+language as small as possible, i.e. to not introduce types or functions
+which are not essential for definitional purposes or essential for providing a practical
+interface to users.
+
+It also discusses some alternatives or additions which should be considered as part of the
+[preliminary investigation](#implementation-plan).
+
+### Providing safe lookups
+
+The `indexArray` builtin function is necessary, since access to constant-time lookup is the main
+requirement outlined in this CIP. However, there remains the question of how users should
+deal with the function's partiality.
+We considered the following options:
+
+1. Introduce another new builtin type, one which implements `Maybe` semantics. The type
+signature for `indexArray` would become `forall a . Integer -> Array a -> Maybe a` and it
+would return `Nothing` for out-of-bounds lookups.
+    - The obvious disadvantage is the necessity of adding another new builtin type (there is
+    no `Maybe` builtin type in Plutus Core), which would further increase the complexity of
+    the builtin language.
+    - Another disadvantage would be that this solution is the most costly: users will incur
+    additional costs in deconstructing the returned `Maybe`.
+
+2. Failed lookups return a default value provided by the caller:
+`indexArray :: forall a . Integer -> a -> Array a -> a`.
+    - This solution is problematic whenever there is no sensible default value and
+    the user wants the function to fail. Since Plutus Core is strict, it is not possible
+    to pass `error` as the default value without it getting evaluated before the call and
+    terminating execution immediately.
+    - As of the time of writing, builtin functions cannot be higher-order. However, that is
+    subject to change in the near future when pattern matching builtins will be supported by
+    Plutus Core. This feature would allow a safe version of `indexArray` with type `forall a .
+    Integer -> (() -> a) -> Array a -> a` to be expressible in the builtins language.
```

Comment:
Sorry, my suggestion had a flaw, the ordering should be like this:

```haskell
indexArray :: (() -> a) -> Array a -> Integer -> a
```

This is so that we can

1. give a name to `indexArray (\_ -> error)` (e.g. `unsafeIndexArray`) and recover the failing version of `indexArray` that way without pretty much any performance overhead (that application will be computed once per execution)
2. give a name to `unsafeIndexArray arr` for a specific `arr`, so that this application also needs to be computed only once

I know it's weird to make `Integer` appear last (although that's kinda like the infix `(!)`), but I think efficiency concerns outweigh beauty ones here, plus compiler authors are free to provide versions of the builtin with arguments flipped.

Oh yeah, and pattern matching builtins have landed, so `indexArray` taking a function should be definable now.